sphinx/themes/basic/static/searchtools.js

```diff
@@ -219,7 +219,16 @@ const Search = {
     (document.body.appendChild(document.createElement("script")).src = url),
 
   setIndex: (index) => {
-    Search._index = index;
+    const sanitiseRecursive = function(obj) {
+      const result = Object.create(null);
+      for (const [k, v] of Object.entries(obj)) {
+          if (!(v instanceof Object)) result[k] = v;
+          else if (Array.isArray(v)) result[k] = Object.freeze(v);
```

Comment:
Hm.  This appears incomplete; our search index format contains some arrays-of-arrays; freezing the outer `Array` does not make the nested arrays within it immutable.

Reply:
(unfortunately that is probably in conflict with the goal of achieving <= 50ms runtime overhead when sanitising the existing Python `searchindex.js`)

Reply:
> Hm. This appears incomplete; our search index format contains some arrays-of-arrays; freezing the outer `Array` does not make the nested arrays within it immutable.

This only occurs in a few places within the index datastructures.  An option in those cases would be to translate the double-nested array representations into a representation that uses an object to represent the outer layer, like so:

```javascript
// original
[[1, 'a'], [2, 'b'], ...]

// restructured
{0: [1, 'a'], 1: [2, 'b'], ...} 
```

This does increase the size of the index file slightly -- but even if+when `records and tuples` (the glorious future that may solve this) are introduced, the syntax for those would necessitate additional characters at the nested-level, I believe.

```javascript
// records+tuples
#[#[1, 'a'], #[2, 'b'], ...]
```

So this isn't quite as outlandish as it might first seem.

Reply:
With either representation, we'll still going to need to freeze the individual `Arrays` contained within the parent array/object, though.  So I'm not sure that this restructuring is worthwhile.

We may as well modify `sanitiseRecursive` to handle nested `Array`s.  I expect the performance cost will be similar, and that would mean we don't need to change the Python code or the search index datastructure.

Reply:
> Next let's try that without the restructuring (compatible/existing search index file format).

That seems better: ~135ms (tested against 0248ccc86f71e69c7b8ad6f69ee6ef0e8518765f).

![image](https://github.com/user-attachments/assets/9bd30eb1-df33-4799-917d-013e1ad37d78)

Now let's try optimizing it.